R/pr.R

```diff
@@ -0,0 +1,556 @@
+#' @param name The name of the cookie in the user's browser.
+#' @param expiration A number representing the number of seconds into the future
+#'   before the cookie expires or a \code{POSIXt} date object of when the cookie expires.
+#'   Defaults to the end of the user's browser session.
+#' @param http Boolean that adds the \code{HttpOnly} cookie flag that tells the browser
+#'   to save the cookie and to NOT send it to client-side scripts. This mitigates \href{https://developer.mozilla.org/en-US/docs/Glossary/Cross-site_scripting}{cross-site scripting}.
+#'   Defaults to \code{TRUE}.
+#' @param secure Boolean that adds the \code{Secure} cookie flag.  This should be set
+#'   when the route is eventually delivered over \href{https://en.wikipedia.org/wiki/HTTPS}{HTTPS}.
+#' @seealso \itemize{
+#' \item \href{https://github.com/jeroen/sodium}{'sodium'}: R bindings to 'libsodium'
+#' \item \href{https://download.libsodium.org/doc/}{'libsodium'}: A Modern and Easy-to-Use Crypto Library
+#' \item \href{https://github.com/r-lib/keyring}{'keyring'}: Access the system credential store from R
+#' \item \href{https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie#Directives}{Set-Cookie flags}: Descriptions of different flags for \code{Set-Cookie}
+#' \item \href{https://developer.mozilla.org/en-US/docs/Glossary/Cross-site_scripting}{Cross-site scripting}: A security exploit which allows an attacker to inject into a website malicious client-side code
+#' }
+#' @examples
+#' \dontrun{
+#'
+#' ## Set secret key using `keyring` (preferred method)
+#' keyring::key_set_with_value("plumber_api", password = plumber::randomCookieKey())
+#'
+#'
+#' pr() %>%
+#'   pr_cookie(
+#'     keyring::key_get("plumber_api"),
+#'     name = "counter"
+#'   ) %>%
+#'   pr_get("/sessionCounter", function(req) {
+#'     count <- 0
+#'     if (!is.null(req$session$counter)){
+#'       count <- as.numeric(req$session$counter)
+#'     }
+#'     req$session$counter <- count + 1
+#'     return(paste0("This is visit #", count))
+#'   }) %>%
+#'   pr_run()
+#'
+#'
+#' #### -------------------------------- ###
+#'
+#'
+#' ## Save key to a local file
+#' pswd_file <- "normal_file.txt"
+#' cat(plumber::randomCookieKey(), file = pswd_file)
+#' # Make file read-only
+#' Sys.chmod(pswd_file, mode = "0600")
+#'
+#' pr() %>%
+#'   pr_cookie(
+#'     readLines(pswd_file, warn = FALSE),
+#'     name = "counter"
+#'   ) %>%
+#'   pr_get("/sessionCounter", function(req) {
+#'     count <- 0
+#'     if (!is.null(req$session$counter)){
+#'       count <- as.numeric(req$session$counter)
+#'     }
+#'     req$session$counter <- count + 1
+#'     return(paste0("This is visit #", count))
+#'   }) %>%
+#'   pr_run()
+#' }
+#' @export
+pr_cookie <- function(pr,
+                      key,
+                      name = "plumber",
+                      expiration = FALSE,
+                      http = TRUE,
+                      secure = FALSE) {
```

Comment:
Add a check like ``if (!inherits(pr, "plumber")) stop("`pr` must be an object of class 'plumber'.")``